airbyte-ci/connectors/pipelines/pipelines/gradle.py

```diff
@@ -121,29 +122,69 @@ async def _run(self) -> StepResult:
+            # Update the persistent gradle cache by resolving all dependencies.
+            # The idea here is to have this persistent cache contain little more than jars and poms.
+            # Also, build the java CDK and publish it to the local maven repository.
+            .with_exec(
+                sh_dash_c(
+                    [
+                        # Ensure that the local maven repository root directory exists.
+                        "mkdir -p /root/.m2",
+                        # Load from the persistent cache.
+                        "(rsync -a --stats /root/gradle-persistent-cache/ /root/.gradle || true)",
+                        # Resolve all dependencies and write their checksums to './gradle/verification-metadata.dryrun.xml'.
+                        f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS)} --write-verification-metadata sha256 help --dry-run",
+                        # Store to the persistent cache.
+                        "(rsync -a --stats /root/.gradle/ /root/gradle-persistent-cache || true)",
+                        # Build the CDK and publish it to the local maven repository.
+                        # Do this last to not pollute the persistent cache.
+                        f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS)} :airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded",
+                    ]
+                )
+            )
+        )
+
+        # Mount only the code needed to build the connector.
+        # This reduces the scope of the inputs to help dagger reuse container layers.
+        # The contents of '/root/.gradle' and '/root/.m2' are by design not overly sensitive to changes in the rest of the git repo.
+        gradle_container = (
+            gradle_container_base
             # TODO: remove this once we finish the project to boost source-postgres CI performance.
             .with_env_variable("CACHEBUSTER", hacks.get_cachebuster(self.context, self.logger))
-            # Mount the gradle cache volume.
-            # We deliberately don't mount it at $GRADLE_HOME, instead we load it there and store it from there using rsync.
-            # This is because the volume is accessed concurrently by all GradleTask instances.
-            # Hence, why we synchronize the writes by setting the `sharing` parameter to LOCKED.
-            .with_mounted_cache("/root/gradle-cache", self.connector_java_build_cache, sharing=CacheSharingMode.LOCKED)
-            # Mount the parts of the repo which interest us in /airbyte.
-            .with_workdir("/airbyte")
+            # Mount the connector-agnostic whitelisted files in the git repo.
             .with_mounted_directory("/airbyte", self.context.get_repo_dir(".", include=include))
+            # Mount the sources for the connector and its dependencies.
             .with_mounted_directory(str(self.context.connector.code_directory), await self.context.get_connector_dir())
```

Comment:
In order to avoid doubling the `mount` operation from host FS to container I think we can get the directory from the   `with_whole_git_repo`:
`.with_mounted_directory("/airbyte", with_whole_git_repo.directory("/airbyte", include=include)`

Reply:
See comment below for why I couldn't do this in the end.